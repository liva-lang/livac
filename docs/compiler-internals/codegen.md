# Code Generation

The code generator emits Rust source code from the IR, handling concurrency transformations, error handling, and runtime integration.

## Location

**File**: `src/codegen.rs` (4,683 lines)

## Overview

The code generator:
1. **Emits Rust Code**: Converts IR → valid Rust syntax
2. **Generates `liva_rt` Module**: Runtime helpers for concurrency and errors
3. **Handles Async/Par**: Transforms concurrency policies into tokio/rayon code
4. **Manages Error Binding**: Implements fallibility system
5. **Tracks Awaitable Tasks**: Lazy await/join optimization

## Architecture

### CodeGenerator

```rust
pub struct CodeGenerator {
    output: String,
    indent_level: usize,
    ctx: DesugarContext,
    in_method: bool,
    in_fallible_function: bool,
    pending_tasks: HashMap<String, TaskInfo>,
    error_binding_vars: HashSet<String>,
    fallible_functions: HashSet<String>,
    // ... other tracking fields
}
```

## Key Responsibilities

### 1. Runtime Module Generation

Generates `liva_rt` module with:

```rust
mod liva_rt {
    use std::future::Future;
    use tokio::task::JoinHandle;
    
    pub struct Error {
        pub message: String,
    }
    
    pub fn spawn_async<F, T>(future: F) -> JoinHandle<T> { ... }
    pub fn fire_async<F>(future: F) { ... }
    pub fn spawn_parallel<F, T>(f: F) -> JoinHandle<T> { ... }
    pub fn fire_parallel<F>(f: F) { ... }
    
    // Display formatting helpers
    pub fn display_array<T: std::fmt::Debug>(arr: &[T]) -> String { ... }
    // ... more helpers
}
```

### 2. Function Generation

**Non-Async, Non-Fallible**:
```rust
fn add(a: i32, b: i32) -> i32 {
    a + b
}
```

**Async**:
```rust
async fn fetchUser(id: i32) -> String {
    let response = reqwest::get(...).await;
    response.text().await.unwrap()
}
```

**Fallible**:
```rust
fn divide(a: i32, b: i32) -> Result<i32, liva_rt::Error> {
    if b == 0 {
        return Err(liva_rt::Error::from("Division by zero"));
    }
    Ok(a / b)
}
```

**Async + Fallible**:
```rust
async fn processData(url: String) -> Result<String, liva_rt::Error> {
    let data = fetch_from_api(&url).await;
    if data.is_empty() {
        return Err(liva_rt::Error::from("No data"));
    }
    Ok(data)
}
```

### 3. Concurrency Transformations

**Async Call (Immediate Await)**:
```liva
let user = async fetchUser(1)
```
↓
```rust
let user = fetchUser(1).await;
```

**Task Async (Deferred Await)**:
```liva
let userTask = task async fetchUser(1)
let user = await userTask
```
↓
```rust
let userTask = liva_rt::spawn_async(async move { fetchUser(1).await });
let user = userTask.await.unwrap();
```

**Parallel Call**:
```liva
let result = par heavyComputation(100)
```
↓
```rust
let result = { heavyComputation(100) };  // Eager evaluation
```

**Task Par**:
```liva
let task = task par compute(100)
let result = await task
```
↓
```rust
let task = liva_rt::spawn_parallel(move || compute(100));
let result = task.await.unwrap();
```

**Fire Async**:
```liva
fire async logEvent("message")
```
↓
```rust
liva_rt::fire_async(async move { logEvent("message").await; });
```

**Fire Par**:
```liva
fire par backgroundCleanup()
```
↓
```rust
liva_rt::fire_parallel(move || { backgroundCleanup(); });
```

### 4. Error Binding

**Basic Error Binding**:
```liva
let result, err = divide(10, 2)
```
↓
```rust
let (result, err) = match divide(10, 2) {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

**Error Binding with Async**:
```liva
let result, err = async divide(10, 2)
```
↓
```rust
let (result, err) = match divide(10, 2).await {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

**Error Binding with Task**:
```liva
let taskHandle = task async divide(10, 2)
let result, err = await taskHandle
```
↓
```rust
let taskHandle = liva_rt::spawn_async(async move { divide(10, 2).await });
let (result, err) = match taskHandle.await.unwrap() {
    Ok(val) => (val, "".to_string()),
    Err(e) => (Default::default(), e.to_string()),
};
```

### 5. Data-Parallel Loops

**Sequential**:
```liva
for item in items {
    process(item)
}
```
↓
```rust
for item in items {
    process(item);
}
```

**Parallel (Rayon)**:
```liva
for par item in items with threads 4 {
    process(item)
}
```
↓
```rust
items.into_par_iter()
    .with_max_threads(4)
    .for_each(|item| {
        process(item);
    });
```

**Vectorized (SIMD)**:
```liva
for vec value in values with simdWidth 4 {
    compute(value)
}
```
↓
```rust
// SIMD generation (simplified)
for chunk in values.chunks_exact(4) {
    compute(chunk[0]);
    compute(chunk[1]);
    compute(chunk[2]);
    compute(chunk[3]);
}
```

**Parallel + Vec**:
```liva
for parvec value in values with simdWidth 4 ordered {
    process(value)
}
```
↓
```rust
values.into_par_iter()
    .with_simd_width(4)
    .ordered()
    .for_each(|value| {
        process(value);
    });
```

### 6. String Templates

```liva
let name = "Alice"
let greeting = $"Hello, {name}!"
```
↓
```rust
let name = "Alice";
let greeting = format!("Hello, {}!", name);
```

Complex templates:
```liva
let msg = $"User {user.name} (age {user.age}) at {timestamp}"
```
↓
```rust
let msg = format!("User {} (age {}) at {}", user.name, user.age, timestamp);
```

### 7. Classes

```liva
Person {
    constructor(name: string, age: number) {
        this.name = name
        this.age = age
    }
    
    name: string
    age: number
    
    greet() => $"Hello, I'm {this.name}"
}
```
↓
```rust
#[derive(Debug, Clone)]
struct Person {
    name: String,
    age: i32,
}

impl Person {
    fn new(name: String, age: i32) -> Self {
        Self { name, age }
    }
    
    fn greet(&self) -> String {
        format!("Hello, I'm {}", self.name)
    }
}
```

## Main Entry Point

```rust
pub fn generate(module: ir::Module, ctx: DesugarContext) -> Result<String>
```

Returns complete Rust source code ready for compilation.

## Code Organization

### String Building

```rust
fn writeln(&mut self, s: &str)  // Write line with indentation
fn indent(&mut self)            // Increase indent
fn dedent(&mut self)            // Decrease indent
```

### Generation Methods

```rust
fn generate_function(&mut self, func: &ir::Function)
fn generate_statement(&mut self, stmt: &ir::Stmt)
fn generate_expression(&mut self, expr: &ir::Expr)
fn generate_for_loop(&mut self, for_stmt: &ir::For)
fn generate_call(&mut self, call: &ir::CallExpr)
```

## Error Codes

| Code | Description |
|------|-------------|
| **E5001** | Unsupported IR construct |
| **E5002** | Invalid concurrency combination |
| **E5003** | Codegen internal error |

## Output Format

Generated Rust code includes:

1. **Module header**: `use` statements for dependencies
2. **`liva_rt` module**: Runtime helpers
3. **Type definitions**: Structs, enums
4. **Functions**: Public and private
5. **Main function**: Entry point (if present)

## Summary

- **4,683 Lines**: Most complex compiler phase
- **IR → Rust**: Direct code emission
- **Runtime Integration**: Generates `liva_rt` module
- **Concurrency Transformation**: Handles async/par/task/fire
- **Error Binding**: Implements fallibility system
- **Data-Parallel**: Rayon and SIMD code generation

**Next**: [Desugaring →](desugaring.md)

**See Also**:
- [Runtime Module](runtime.md)
- [IR](ir.md)
- [Concurrency](../language-reference/concurrency.md)
