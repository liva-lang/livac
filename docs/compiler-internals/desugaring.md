# Desugaring

The desugaring pass collects metadata about the program and prepares it for code generation.

## Location

**File**: `src/desugaring.rs` (260 lines)

## Overview

Desugaring is a **pre-codegen analysis pass** that:
1. **Collects Dependencies**: Tracks `use rust` declarations
2. **Detects Concurrency**: Determines if tokio/rayon needed
3. **Builds Context**: Creates `DesugarContext` for codegen

## DesugarContext

```rust
pub struct DesugarContext {
    pub rust_crates: Vec<(String, Option<String>)>,  // (crate, alias)
    pub has_async: bool,
    pub has_parallel: bool,
}
```

This context tells codegen what runtime dependencies to include.

## Main Function

```rust
pub fn desugar(program: Program) -> Result<DesugarContext>
```

**Returns**: Metadata about program's runtime needs.

## Process

### 1. Collect `use rust` Declarations

```liva
use rust reqwest
use rust serde as json
```

Collected as:
```rust
rust_crates: [
    ("reqwest".to_string(), None),
    ("serde".to_string(), Some("json".to_string())),
]
```

### 2. Detect Async Usage

Checks for:
- `async` calls: `async fetchData()`
- Async functions: Functions marked `is_async_inferred`
- Async policies: `ExecPolicy::Async`, `TaskAsync`, `FireAsync`

If found:
```rust
ctx.has_async = true
rust_crates.push(("tokio".to_string(), None))  // Auto-add tokio
```

### 3. Detect Parallel Usage

Checks for:
- `par` calls: `par compute()`
- Parallel loops: `for par x in items`
- Parallel policies: `ExecPolicy::Par`, `TaskPar`, `FirePar`
- Data-parallel: `DataParallelPolicy::Par`, `Vec`, `ParVec`

If found:
```rust
ctx.has_parallel = true
// Rayon included automatically in liva_rt
```

## Example

**Input Program**:
```liva
use rust reqwest

fetchUser(id: number): string {
  let response = async httpGet($"/users/{id}")
  return response.body
}

processData(items) {
  for par item in items with threads 4 {
    compute(item)
  }
}

main() {
  let user = fetchUser(1)
  processData([1, 2, 3, 4])
}
```

**Output Context**:
```rust
DesugarContext {
    rust_crates: [
        ("reqwest", None),
        ("tokio", None),  // Auto-added
    ],
    has_async: true,     // fetchUser has async call
    has_parallel: true,  // processData has par loop
}
```

## Concurrency Detection

### Traversal Functions

```rust
fn check_concurrency(item: &TopLevel, ctx: &mut DesugarContext)
fn check_block_concurrency(body: &IfBody, ctx: &mut DesugarContext)
fn check_stmt_concurrency(stmt: &Stmt, ctx: &mut DesugarContext)
fn check_expr_concurrency(expr: &Expr, ctx: &mut DesugarContext)
```

**Recursively walks AST** to find concurrency usage.

### Async Detection

```rust
match call.exec_policy {
    ExecPolicy::Async | ExecPolicy::TaskAsync | ExecPolicy::FireAsync => {
        ctx.has_async = true;
    }
    _ => {}
}
```

### Parallel Detection

```rust
match for_stmt.policy {
    DataParallelPolicy::Par | DataParallelPolicy::Vec | DataParallelPolicy::ParVec => {
        ctx.has_parallel = true;
    }
    _ => {}
}
```

## Codegen Integration

Codegen uses `DesugarContext` to:

1. **Include Dependencies**:
   ```rust
   use tokio;
   use rayon::prelude::*;
   ```

2. **Generate Runtime Module**:
   ```rust
   mod liva_rt {
       // Include async helpers if ctx.has_async
       // Include parallel helpers if ctx.has_parallel
   }
   ```

3. **Wrap Main**:
   ```rust
   #[tokio::main]  // If ctx.has_async
   async fn main() {
       // ...
   }
   ```

## Why "Desugaring"?

The name is **legacy** from earlier designs where this pass did more transformation. Now it's primarily **metadata collection**.

**Better name**: `MetadataCollector` or `ContextBuilder`

## Summary

- **260 Lines**: Lightweight analysis pass
- **Metadata Collection**: Determines runtime needs
- **Dependency Tracking**: Manages `use rust` crates
- **Concurrency Detection**: Identifies async/parallel usage
- **Codegen Preparation**: Builds context for code generation

**Next**: [Runtime Module â†’](runtime.md)

**See Also**:
- [Code Generation](codegen.md)
- [Architecture](architecture.md)
