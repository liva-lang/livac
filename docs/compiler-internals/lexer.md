# Lexer (Tokenization)

The lexer converts source code into tokens using the **Logos** crate for high-performance lexical analysis.

## Location

**File**: `src/lexer.rs`

## Overview

The lexer performs:
1. **Tokenization**: Converts character stream to tokens
2. **Comment Stripping**: Removes single-line (`//`) and multi-line (`/* */`) comments
3. **Whitespace Handling**: Skips whitespace automatically
4. **Span Tracking**: Records source locations for error reporting

## Token Types

### Keywords (59 tokens)

**Control Flow**:
- `let`, `const`, `if`, `else`, `while`, `for`, `in`, `switch`, `case`, `default`
- `return`, `throw`, `try`, `catch`

**Concurrency**:
- `async`, `parallel`, `par`, `task`, `fire`, `await`
- `move`, `seq`, `vec`, `parvec`, `with`

**Data-Parallel Policies**:
- `ordered`, `chunk`, `threads`, `simdWidth`
- `prefetch`, `reduction`, `schedule`, `detect`
- `auto`, `safe`, `fast`, `static`, `dynamic`

**Error Handling**:
- `fail`

**Types**:
- `number`, `float`, `bool`, `char`, `string`, `bytes`
- `type`, `import`, `use`, `rust`, `test`

**Literals**:
- `true`, `false`

**Operators** (word-based):
- `and`, `or`, `not`, `as`

### Operators

**Arithmetic**: `+`, `-`, `*`, `/`, `%`

**Comparison**: `==`, `!=`, `<`, `<=`, `>`, `>=`

**Logical**: `&&`, `||`, `!`

**Other**: `=`, `.`, `..`, `=>`, `?`, `:`

### Delimiters

`(`, `)`, `{`, `}`, `[`, `]`, `,`, `;`

### Literals

```rust
IntLiteral(i64)         // 42, 100, 1_000_000
FloatLiteral(f64)       // 3.14, 1.5, 0.001
StringLiteral(String)   // "hello", "world"
CharLiteral(char)       // 'a', 'Z', '\n'
StringTemplate(String)  // $"Hello {name}"
```

### Identifiers

```rust
Ident(String)           // public: myVar, calculate
ProtectedIdent(String)  // protected: _helper
PrivateIdent(String)    // private: __secret
```

Visibility is determined by leading underscores:
- No prefix: Public
- `_`: Protected
- `__`: Private

## Implementation

### Using Logos

```rust
#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t\r\n\f]+")]         // Skip whitespace
#[logos(skip r"//[^\n]*")]             // Skip single-line comments
#[logos(skip r"/\*([^*]|\*[^/])*\*/")]  // Skip multi-line comments
pub enum Token {
    #[token("let")]
    Let,
    
    #[token("async")]
    Async,
    
    #[regex(r"[0-9][0-9_]*", parse_int)]
    IntLiteral(i64),
    
    // ... more tokens
}
```

### Key Features

1. **Automatic Whitespace Skipping**: Using `#[logos(skip)]`
2. **Regex Patterns**: For numbers, strings, identifiers
3. **Underscore Separators**: `1_000_000` → `1000000`
4. **String Templates**: `$"Hello {name}"` tokenized separately

### Integer Parsing

```rust
#[regex(r"[0-9][0-9_]*", |lex| {
    lex.slice().replace('_', "").parse::<i64>().ok()
})]
IntLiteral(i64),
```

Supports underscores for readability: `1_000_000`

### Float Parsing

```rust
#[regex(r"[0-9][0-9_]*\.[0-9][0-9_]*", |lex| {
    lex.slice().replace('_', "").parse::<f64>().ok()
})]
FloatLiteral(f64),
```

### String Literals

```rust
#[regex(r#""([^"\\]|\\.)*""#, |lex| {
    let s = lex.slice();
    let content = &s[1..s.len()-1];  // Strip quotes
    Some(content.to_string())
})]
StringLiteral(String),
```

### String Templates

```rust
#[regex(r#"\$"([^"\\]|\\.)*""#, |lex| {
    let s = lex.slice();
    let content = &s[2..s.len()-1];  // Strip $" and "
    Some(content.to_string())
})]
StringTemplate(String),
```

## TokenWithSpan

```rust
pub struct TokenWithSpan {
    pub token: Token,
    pub span: Span,
}
```

Tracks source location for error reporting:
- `span.start`: Start byte offset
- `span.end`: End byte offset
- `span.start_position(map)`: (line, column)

## Public API

```rust
// Tokenize source code
pub fn tokenize(source: &str, file_name: &str) -> Result<Vec<TokenWithSpan>>
```

Returns `Vec<TokenWithSpan>` or `CompilerError` on invalid tokens.

## Error Handling

Lexer errors include:
- **Invalid tokens**: Unrecognized characters
- **Malformed literals**: Unclosed strings, invalid numbers
- **Location info**: Line, column, source snippet

## Summary

- **50+ Keywords**: Including all concurrency and data-parallel policy tokens
- **Logos-based**: High-performance lexing
- **Span Tracking**: For precise error messages
- **Comment Stripping**: Automatic via regex
- **Visibility-Aware**: Identifier prefixes determine visibility

**Next**: [Parser →](parser.md)
