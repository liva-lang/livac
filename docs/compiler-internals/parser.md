# Parser (Syntax Analysis)

The parser converts tokens into an **Abstract Syntax Tree (AST)** using recursive descent parsing.

## Location

**File**: `src/parser.rs` (1,754 lines)

## Overview

The parser performs:
1. **Token Stream Processing**: Consumes tokens from lexer
2. **AST Construction**: Builds typed AST nodes
3. **Syntax Validation**: Checks grammar rules
4. **Error Recovery**: Reports syntax errors with context

## Architecture

### Recursive Descent Parser

```rust
pub struct Parser {
    tokens: Vec<TokenWithSpan>,
    current: usize,
    source_file: String,
    source_code: String,
    source_map: SourceMap,
}
```

**Key Characteristics**:
- **Top-down parsing**: Starts from `Program` → `TopLevel` → ... → `Expr`
- **Predictive**: Looks ahead (`peek()`) to decide parsing path
- **Hand-written**: Not generated from grammar (flexibility for error recovery)

## Parsing Entry Points

### Main Entry

```rust
pub fn parse_program(&mut self) -> Result<Program>
```

Returns `Program { items: Vec<TopLevel> }`

### Top-Level Items

```rust
fn parse_top_level(&mut self) -> Result<TopLevel>
```

Parses:
- `import` declarations
- `use rust` declarations
- `type` definitions
- `class` definitions
- Function declarations
- `test` declarations

## Key Parsing Functions

### Functions

```rust
fn parse_function(&mut self, name: String) -> Result<FunctionDecl>
```

Handles:
- Type parameters: `<T, U>`
- Parameters with types and defaults
- Return type annotations
- Block body or arrow expression body

### Classes

```rust
fn parse_class(&mut self) -> Result<ClassDecl>
```

Features:
- Inheritance: `ChildClass: ParentClass`
- Constructor with parameters
- Fields with visibility
- Methods (block or arrow)

### Statements

```rust
fn parse_statement(&mut self) -> Result<Stmt>
```

Statement types:
- **Variable**: `let x = 10` or `let x, err = fallible()`
  - **Destructuring** (v0.10.2+): `let {name, age} = user` or `let [first, second] = array`
- **Constant**: `const MAX = 100`
- **If**: With optional parentheses, single-statement or block
- **While**: `while condition { ... }`
- **For**: With data-parallel policies (`seq`, `par`, `vec`, `parvec`)
- **Switch**: Case-based branching
- **Try-Catch**: Exception handling
- **Return**: `return value`
- **Throw**: `throw error`
- **Fail**: `fail "message"`
- **Expression**: Any expression as statement

### Variable Bindings and Destructuring (v0.10.2+)

```rust
fn parse_let_bindings(&mut self) -> Result<Vec<VarBinding>>
```

Liva supports multiple binding patterns:

**Simple Binding**:
```liva
let name = "Alice"
let age: number = 25
```

**Fallible Binding** (error tuple):
```liva
let result, err = parseInt("123")
let data, error = HTTP.get("https://api.example.com")
```

**Object Destructuring** (v0.10.2+):
```liva
let {name, age} = user
let {name: userName, age: userAge} = user  // with rename
```

**Array Destructuring** (v0.10.2+):
```liva
let [first, second] = array
let [first, , third] = array  // skip elements
let [head, ...tail] = array   // rest pattern
```

**AST Structure**:

```rust
pub struct VarBinding {
    pub pattern: BindingPattern,
    pub type_ref: Option<TypeRef>,
    pub span: Span,
}

pub enum BindingPattern {
    Identifier(String),
    Object(ObjectPattern),
    Array(ArrayPattern),
}

pub struct ObjectPattern {
    pub fields: Vec<ObjectPatternField>,
}

pub struct ObjectPatternField {
    pub key: String,
    pub binding: String,  // renamed binding or same as key
}

pub struct ArrayPattern {
    pub elements: Vec<Option<String>>,  // None for skipped elements
    pub rest: Option<String>,           // rest pattern variable name
}
```

**Helper Methods**:
- `binding.name() -> Option<&str>`: Returns the simple identifier name, or None for complex patterns
- `binding.is_simple() -> bool`: Returns true if it's a simple identifier binding

### For Loops with Policies

```rust
// Parses: for <policy> var in iterable with <options> { body }
if self.match_token(&Token::For) {
    let mut policy = DataParallelPolicy::Seq;
    
    // Parse policy keyword (seq, par, vec, parvec)
    if self.match_token(&Token::Seq) {
        policy = DataParallelPolicy::Seq;
    } else if self.match_token(&Token::Par) {
        policy = DataParallelPolicy::Par;
    } // ... etc
    
    let var = self.parse_identifier()?;
    self.expect(Token::In)?;
    let iterable = self.parse_expression()?;
    
    // Parse with options
    let options = if self.match_token(&Token::With) {
        self.parse_for_options()?
    } else {
        ForPolicyOptions::default()
    };
    
    // Parse body
    self.expect(Token::LBrace)?;
    let body = self.parse_block_stmt()?;
    self.expect(Token::RBrace)?;
    
    return Ok(Stmt::For(ForStmt { var, iterable, policy, options, body }));
}
```

### Expressions (Operator Precedence)

```rust
fn parse_expression(&mut self) -> Result<Expr>
fn parse_ternary(&mut self) -> Result<Expr>
fn parse_logical_or(&mut self) -> Result<Expr>
fn parse_logical_and(&mut self) -> Result<Expr>
fn parse_equality(&mut self) -> Result<Expr>
fn parse_comparison(&mut self) -> Result<Expr>
fn parse_range(&mut self) -> Result<Expr>
fn parse_additive(&mut self) -> Result<Expr>
fn parse_multiplicative(&mut self) -> Result<Expr>
fn parse_unary(&mut self) -> Result<Expr>
fn parse_postfix(&mut self) -> Result<Expr>
fn parse_primary(&mut self) -> Result<Expr>
```

**Precedence Climbing**: Each function handles one precedence level.

### Call Expressions

```rust
fn parse_postfix(&mut self) -> Result<Expr>
```

Handles:
- **Simple calls**: `func(args)`
- **Concurrency policies**: `async func(args)`, `par func(args)`
- **Task creation**: `task async func(args)`
- **Fire-and-forget**: `fire par func(args)`
- **Member access**: `obj.property`
- **Index access**: `arr[index]`
- **Method calls**: `obj.method(args)`

### Primary Expressions

```rust
fn parse_primary(&mut self) -> Result<Expr>
```

Parses:
- **Literals**: Numbers, strings, booleans, chars
- **Identifiers**: Variable names
- **Parenthesized**: `(expr)`
- **Array literals**: `[1, 2, 3]`
- **Object literals**: `{ name: "Alice", age: 25 }`
- **Struct literals**: `Person { name: "Alice", age: 25 }`
- **String templates**: `$"Hello {name}"`
- **Lambdas**: `(x, y) => x + y` or `move (x) => { ... }`

### Lambda Parsing

```rust
fn parse_lambda_expr(&mut self) -> Result<LambdaExpr>
```

Features:
- **Move capture**: `move (x) => x * 2`
- **Parameter types**: `(x: number, y: number) => x + y`
- **Return types**: `(x: number): number => x * 2`
- **Block bodies**: `(x) => { return x * 2 }`
- **Expression bodies**: `(x) => x * 2`

## Error Handling

### Error Recovery

```rust
fn error(&self, message: String) -> CompilerError
```

Generates rich error messages with:
- **Line and column**: Precise location
- **Source snippet**: Shows problematic code
- **Error code**: E2xxx codes for parser errors
- **Helpful message**: Explains what went wrong

### Example Error

```
[E2010] Expected ';' after statement
  --> main.liva:15:20
   |
15 |     let x = 10
   |               ^ Expected ';' here
```

## AST Structure

### Top-Level Items

```rust
pub enum TopLevel {
    Import(ImportDecl),
    UseRust(UseRustDecl),
    Type(TypeDecl),
    Class(ClassDecl),
    Function(FunctionDecl),
    Test(TestDecl),
}
```

### Statements

```rust
pub enum Stmt {
    VarDecl(VarDecl),
    ConstDecl(ConstDecl),
    Assign(AssignStmt),
    If(IfStmt),
    While(WhileStmt),
    For(ForStmt),
    Switch(SwitchStmt),
    TryCatch(TryCatchStmt),
    Throw(ThrowStmt),
    Fail(FailStmt),
    Return(ReturnStmt),
    Expr(ExprStmt),
    Block(BlockStmt),
}
```

### Expressions

```rust
pub enum Expr {
    Literal(Literal),
    Identifier(String),
    Binary { op: BinOp, left: Box<Expr>, right: Box<Expr> },
    Unary { op: UnOp, operand: Box<Expr> },
    Ternary { condition, then_expr, else_expr },
    Call(CallExpr),
    Member { object: Box<Expr>, property: String },
    Index { object: Box<Expr>, index: Box<Expr> },
    ObjectLiteral(Vec<(String, Expr)>),
    StructLiteral { type_name: String, fields: Vec<(String, Expr)> },
    ArrayLiteral(Vec<Expr>),
    Lambda(LambdaExpr),
    StringTemplate { parts: Vec<StringTemplatePart> },
    Fail(Box<Expr>),
}
```

### Call Expression (with Policies)

```rust
pub struct CallExpr {
    pub callee: Box<Expr>,
    pub args: Vec<Expr>,
    pub exec_policy: ExecPolicy,
}

pub enum ExecPolicy {
    Normal,
    Async,
    Par,
    TaskAsync,
    TaskPar,
    FireAsync,
    FirePar,
}
```

### For Loop Policies

```rust
pub enum DataParallelPolicy {
    Seq,
    Par,
    Vec,
    ParVec,
}

pub struct ForPolicyOptions {
    pub ordered: bool,
    pub chunk: Option<i64>,
    pub threads: Option<ThreadOption>,
    pub simd_width: Option<SimdWidthOption>,
    pub prefetch: Option<i64>,
    pub reduction: Option<ReductionOption>,
    pub schedule: Option<ScheduleOption>,
    pub detect: Option<DetectOption>,
}
```

## Public API

```rust
pub fn parse(tokens: Vec<TokenWithSpan>, file_name: &str, source_code: &str) 
    -> Result<Program>
```

## Summary

- **Recursive Descent**: Hand-written, predictive parser
- **Rich AST**: Fully typed nodes with source spans
- **Policy-Aware**: Parses concurrency and data-parallel annotations
- **Error Recovery**: Detailed error messages with source context
- **1,754 Lines**: Comprehensive grammar support

**Next**: [Semantic Analysis →](semantic.md)

**See Also**:
- [AST Reference](../language-reference/syntax-overview.md)
- [Grammar](grammar.md)
