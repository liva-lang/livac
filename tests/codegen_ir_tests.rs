use insta::assert_snapshot;
use livac::codegen::generate_from_ir;
use livac::desugaring::desugar;
use livac::lexer::tokenize;
use livac::lowering::lower_program;
use livac::parser::parse;
use livac::semantic::analyze;

fn compile_ir(source: &str) -> (String, String) {
    let tokens = tokenize(source).expect("tokenize");
    let program = parse(tokens, source).expect("parse");
    let analyzed = analyze(program.clone()).expect("semantic");
    let ctx = desugar(analyzed.clone()).expect("desugar");
    let module = lower_program(&analyzed);

    generate_from_ir(&module, &program, ctx).expect("codegen")
}

#[test]
fn ir_codegen_simple_function() {
    let source = r#"
        add(a: number, b: number): number = a + b

        main() {
            let result = add(2, 3)
            print(result)
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);
    assert_snapshot!("ir_simple_function", rust_code);
}

#[test]
fn ir_codegen_async_helpers() {
    let source = r#"
        fetch() {
            let response = async getData()
            return response
        }

        main() {
            let value = async fetch()
            fire async fetch()
            print(value)
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);
    assert_snapshot!("ir_async_helpers", rust_code);
}

#[test]
fn ir_codegen_parallel_helpers() {
    let source = r#"
        heavy() {
            return parallel compute()
        }

        main() {
            let value = parallel heavy()
            fire parallel heavy()
            print(value)
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);
    assert_snapshot!("ir_parallel_helpers", rust_code);
}

#[test]
fn ir_codegen_try_catch_and_switch() {
    let source = r#"
        main() {
            try {
                let flag = false
                if flag == false {
                    throw "error"
                }
                print("done")
            } catch (err) {
                print(err)
            }

            switch 200 {
                case 200:
                    print("success")
                case 500:
                    print("server")
                default:
                    print("other")
            }
        }
    "#;

    let (rust_code, _cargo) = compile_ir(source);
    assert_snapshot!("ir_try_catch_switch", rust_code);
}
