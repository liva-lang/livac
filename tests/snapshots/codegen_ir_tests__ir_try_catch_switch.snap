---
source: tests/codegen_ir_tests.rs
assertion_line: 97
expression: rust_code
---
fn main() {
    match (|| -> Result<(), Box<dyn std::error::Error>> {
        let flag = false;
        if flag == false {
            return Err("error".into());
        }
        println!("{}", "done");
        Ok(())
    })() {
        Ok(_) => {},
        Err(err) => {
            println!("{}", err);
        }
    }
    match 200 {
        200 => {
            println!("{}", "success");
        },
        500 => {
            println!("{}", "server");
        },
        _ => {
            println!("{}", "other");
        },
    }
}
